config/schemas/config.yaml

```diff
@@ -126,6 +411,654 @@ properties:
           - containerd
           - docker
     additionalProperties: false
+  calicoAccountant:
+    additionalProperties: false
+    properties:
+      backend:
+        default: iptables
+        enum:
+          - iptables
+          - nftables
+        type: string
+      enabled:
+        default: true
+        type: boolean
+      resources:
+        $ref: '#/$defs/resources'
+      tolerations:
+        $ref: '#/$defs/tolerations'
+    type: object
+  calicoFelixMetrics:
+    additionalProperties: false
+    properties:
+      enabled:
+        default: true
+        type: boolean
+    type: object
+  certmanager:
+    additionalProperties: false
+    properties:
+      affinity:
+        $ref: '#/$defs/affinity'
+      cainjector:
+        $ref: '#/$defs/commonResource'
+      extraArgs:
+        $ref: '#/$defs/extraArgs'
+      nodeSelector:
+        $ref: '#/$defs/nodeSelector'
+      resources:
+        $ref: '#/$defs/resources'
+      tolerations:
+        $ref: '#/$defs/tolerations'
+      topologySpreadConstraints:
+        $ref: '#/$defs/topologySpreadConstraints'
+      webhook:
+        $ref: '#/$defs/commonResource'
+    type: object
+  ingressNginx:
+    additionalProperties: false
+    properties:
+      controller:
+        additionalProperties: false
+        properties:
+          additionalConfig:
+            type: object
+          affinity:
+            $ref: '#/$defs/affinity'
+          allowSnippetAnnotations:
+            default: false
+            type: boolean
+          chroot:
+            type: boolean
+          config:
+            additionalProperties: false
+            properties:
+              useProxyProtocol:
+                type: boolean
+              annotationsRiskLevel:
+                type: string
+                description: See [reference](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#annotations-risk-level)
+                enum:
+                  - Critical
+                  - High
+                  - Medium
+                  - Low
+            type: object
+          enablepublishService:
+            default: false
+            type: boolean
+          enableAnnotationValidations:
+            description: |-
+              Enable validation of Nginx annotations on Ingress resources.
+              This is kept false by default due to the maturity of the feature,
+              e.g. lack of documentation and disruptive bugs.
+            default: false
+            type: boolean
+          extraArgs: true
+          nodeSelector:
+            $ref: '#/$defs/nodeSelector'
+          resources:
+            $ref: '#/$defs/resources'
+          service:
+            additionalProperties: false
```

Comment:
Is there some way to avoid validating part of a schema based  on the value of another variable? Because when I tested I had `service.enabled=false` and  thus the `service.type` is not set (or rather it is set to `set-me`) and it fails the schema validation:

```console
[ck8s] Validating wc config
[ck8s] Failed schema validation:
wc-config.yaml: ingressNginx.controller.service.type: ingressNginx.controller.service.type must be one of the following: "ClusterIP", "NodePort", "LoadBalancer", "ExternalName"
```

Reply:
Yes, it's possible but more complex, using either `oneOf` or `if-then` constructs.

Do we want to keep the unused `set-me` entries? Deleting them from the configs when not used would make it less cluttered, and easier to validate if unused settings aren't present at all.

Example `oneOf` method:


```yaml
oneOf:
- properties:
    enabled:
      const: false
- properties:
    enabled:
      const: true
  # other properties
  required: [ type, ... ] # list of mandatory properties if enabled
```

Reply:
I am not sure what would be best in this case, removing unused `set-me` entries would require some reworking of our current config and scripts. Maybe improving both with e.g. adding `oneOf` for templating which we have `enabled` for and rework `set-me`s a bit in a follow up task. Maybe someone else has an opinion on this?

Reply:
I think I'd prefer if, then, else for boolean flags given that it becomes very explicit.

And the false case could be allow anything for now, and I think we need to take a decision on how we want to handle optional configuration with this moving forward. But that will become more important once we have a way to render the default configuration from the schema.

Reply:
Turning it into an `if-then` then 😄